runtime/lua/vim/lsp/completion.lua

```diff
@@ -0,0 +1,93 @@
+local M = {}
+
+--- @class vim.lsp.completion.BufHandle
+--- @field clients table<integer, lsp.Client>
+--- @field triggers table<string, lsp.Client[]>
+
+--- @type table<integer, vim.lsp.completion.BufHandle>
+local buf_handles = {}
+
+local function complete_done() end
+
+--- Initializes the completion commands for the given client.
+---
+--- @param client lsp.Client
+local function init_commands(client)
+  local trigger_completion_cmd = 'editor.action.triggerSuggest'
+
+  -- Check if the command is in the global registry or in the client's commands.
+  if
+    not vim.lsp.commands[trigger_completion_cmd] and not client.commands[trigger_completion_cmd]
+  then
+    client.commands[trigger_completion_cmd] = function()
+      local ok, res = pcall(M.trigger_completion)
+      return ok and vim.NIL
+        or vim.lsp.rpc_response_error(vim.lsp.protocol.ErrorCodes.InternalError, res)
+    end
+  end
+end
+
+function M.trigger_completion() end
+
+--- TODO(mariasolos): Document.
+---
+--- @param client_id integer Client ID
+--- @param bufnr integer Buffer handle, or 0 for the current buffer
+function M.attach(client_id, bufnr)
```

Comment:
I'm very skeptical about adding "attach to client" (or "attach to completion" is an even deeper _inner platform_) as yet another way to extend the LSP client. There are already so many ways: start, callbacks, protocol handlers (local and global).

> Make it opt-in via a `vim.lsp.completion.attach(client, bufnr, [opts])` method

If we enabled it by default, then it could be _opt-out_, by documenting how to delete the `CompleteDone` handler. Any problem with that?

Reply:
I think I prefer opt-in, at least while we develop this feature and make sure it has everything we want/need of core LSP completion.

Reply:
I've made the feature enabled by default, documenting how to disable it if desired. Lmk what you think @justinmk :)

Reply:
Okay I haven't added the disabling part, gimme a minute

Reply:
After further discussion, I think that making the feature opt-in is better for now. That will facilitate passing options (see [Mathias' comment](https://github.com/neovim/neovim/pull/27339#discussion_r1515698767)).

Reply:
> Reluctant to add "attach to client" (or "attach to completion") as yet another way to extend the LSP client. There are already so many ways: start, callbacks, protocol handlers (local and global).

Was that resolved? Seems like this PR is still adding a new "completion provider" concept.

Reply:
@justinmk I thought we decided that opt-in behavior was better for now. Quoting @mfussenegger:
> If we go for opt-out instead of opt-in it also opens the question of how to configure the options (trigger characters, later on maybe the debounce interval, if backspace should re-trigger, and maybe server-side fuzzy search support)

We were also unsure about how to deal with completion and snippet keymaps, but I guess that's no longer relevant with the newly added defaults.

Reply:
> we decided that opt-in behavior was better for now. 

But does that require exposing all these different functions, which also implies this "attach a completion provider" concept? 

My suggestion for opt-in was:

> Alternatively, for _opt-in_ behavior, we could document how to define a `CompleteDone` handler that calls a function we provide.

are there problems with this approach? I would expect this to only require exposing 1 function. And no attach/detach lifecycle.

Reply:
I won't comment on the specific API, but don't forget that the goal here is for plugins to be able to hook into that -- both completion and snippets. 

Also, we won't speedrun this into 0.10; this is getting the pieces into shape for a concerted "LSP autocompletion with snippets" early in the 0.11 cycle so we can iterate on the full picture.

Reply:
(and, yes, if we can trigger the side effects robustly through a `CompleteDone` handler, that would be the preferred way of course; contingent on #20671)

Reply:
It's currently `InsertCharPre`, `InsertLeave`, `TextChangedP`, `TextChangedI` and `CompleteDone`.
(Although `TextChangedP` / `TextChangedI` might disappear, see comment below)

And there are two independent aspects here: auto-trigger, and snippet expansion/side-effects.

Snippet expansion should also be useable with the omnifunc triggered manually, and with having opted out of auto-trigger. So we'd need a way to control these two aspects individually.

Exposing the group name or autocmd ids feels a bit like exposing the implementation details, and might limit how we can extend the implementation later.


I'd suggest to go with the `enable` / `is_enabled`.
That also gives us the option to add configuration options to the `enable` call.

Or maybe I completely misunderstand the proposal here?

Reply:
> (and, yes, if we can trigger the side effects robustly through a `CompleteDone` autocommand, that would be the preferred way of course; contingent on #20671 and a good story how plugins can remove or override such default handlers)

Which we now can \o/ @MariaSolOs rebase on `master`?